### ISSUE TITLE
compiler slow down when compiling many string literals

### ISSUE THREAD

**Issue by gadget114514** [yasl-lang/yasl#124]
on v0.3.6, 
in order to reproduce hung up, the procedure is shown below. 
1. the following code is saved as a.rb; then ruby a.rb > test.yasl
2. ./YASL test.yasl  -> hung up
3. if N is changed to 10, it works well.
```
puts("val := {};");
i = 1;
N=100;
while i < N  do
	puts("fn f#{i}() {");
	puts("echo \"#{i}abcdefghijklmnopqrstu#{i}abcdefghijiklmlnopquskalfe#{i}\";");
	puts("val[\"c#{i+1}\"] = 1;");
	puts("}");
	i = i+1;
end
i = 1;
puts("f1();");
while i < N  do
	puts("if val[\"c#{i}\"] == 1 {");
	puts("f#{i}();");
	puts("}");
	i = i+1;
end
```

**Comment by CoffeeTableEspresso** [comment]
yup this looks like a bug, I'll look into this.

Thanks for reporting this!

**Comment by CoffeeTableEspresso** [comment]
I've looked into it, turns out it's not stuck, it's just taking a very long time to compile. (I left your script alone for a while and came back to it being done.) I'm looking into why this is because this is obviously too slow to be usable.

**Comment by gadget114514** [comment]
i see. it tasks 6 minutes on my bash on windows. 
it looks taking huge trial to find unused slot in hashtable.

**Comment by CoffeeTableEspresso** [comment]
I've run the profiler and it looks like 66% of the time is spent in `hash_function`, which seems to indicate for each `table_search` and `table_insert` we're calling `hash_function` way too many times. (i.e. there's probably too many collisions in the hash function for string, so we're having trouble finding an open spot.)

**Comment by CoffeeTableEspresso** [comment]
it turns out that `f17` is getting repeatedly hashed to the same value which is causing the problem. I'm going to fix the double hashing so we don't have this.

**Comment by CoffeeTableEspresso** [comment]
ok should be fixed now (and this shouldn't happen in the future since the hash function should be fixed now)

**Comment by CoffeeTableEspresso** [comment]
I've merged the fix into the v0.3.7 branch. Your script seems to be good, along with a few others I created that also had this issue.
